README.md
Shopify API
=
This is a simple PHP library that provides a quick and easy way to work with Shopify REST API.
It uses Guzzle as HTTP client. 

IMPORTANT
=
This branch created from v0.2.0 release to support some old projects that are still on the old version of PHP. 
It's highly recommended to use PHP 7.1+ and the latest release of this library. 
I'm not going to support this branch too long or use versioning here. Use it at your own risk. 

Installation
-

```shell
composer require vladimircatrici/shopify
```

Usage
-

### Initialization
```php
require_once 'vendor/autoload.php';
use VladimirCatrici\Shopify;
$clientConfig = new ClientConfig([
    'handle' => 'your-shop-handle', // with or without ".myshopify.com"
    'accessToken' => 'your-access-token'
]);
Shopify\ClientManager::setConfig('default', $clientConfig);
$api = Shopify\ClientManager::get('default');
```

### Configuration

There are a few additional options you can pass to the ClientManager.

-   `apiVersion` (default: `the oldest stable supported version`)
The Shopify API version you want to use.  
Read more about [API versioning at Shopify](https://help.shopify.com/en/api/versioning).   

-   `maxAttemptsOnServerErrors` (default: `1`)  
Number of attempts trying to execute the request. 
It's useful because sometimes Shopify may respond with 500 error.
I would recommend set this to `2` or `3`. The default value is `1` though. 

-   `maxAttemptsOnRateLimitErrors` (default: `1`)  
Number of attempts trying to execute the request on getting `429 Too Many Connections` error.
This might be useful if the same API key is used by other apps which may lead to exceeding the rate limit.
The recommended value would be somewhere under the 10.  

-   `maxLimitRate` (default: `0.5`)  
Number between 0 and 1 describing the maximum limit rate the client should reach before going sleep. 
See `max_limit_rate_sleep_sec` option.  

-   `maxLimitRateSleepSeconds` (default: `1`)  
Number of seconds to sleep when API reaches the maximum API limit rate specified in `max_limit_rate` option.

-   `responseFormatter` (default: instance of `ResponseDefaultFormatter`)  
You can create and set up another data formatter for response you're getting from Shopify.
Your formatter should implement simple `ResponseDataFormatterInterface`. 

### Basic usage

The client implements all 4 HTTP methods that Shopify REST API supports. Method names are:
-   get(_string_ $endpoint, _array_ $options)
-   post(_string_ $endpoint, _array_ $data)
-   put(_string_ $endpoint, _array_ $data)
-   delete(_string_ $endpoint)

The `$endpoint` parameter must always be a string that represents a Shopify API endpoint. 
It should not contain the `/admin/api/#{api_version}/` part in the beginning. 
The `.json` is not necessary in the end of the path as well. For example if Shopify documentations 
shows the endpoint path as `GET /admin/api/#{api_version}/orders.json`, you can just use:
```php
$api->get('orders');
```
 
See more examples below:

#### Get items

```php
$numProducts = $api->get('products/count'); // int
$products = $api->get('products'); // array
foreach ($products as $product) {
  echo sprintf('#%d. %s<br>', 
    $product['id'], $product['title']);
}
```

#### Get `id` and `title` fields of 250 items from the 2nd page

```php
$products = $api->get('products', [
  'fields' => 'id,title',
  'limit' => 250,
  'page' => 2
]);
```

#### Get single item
```php
$product = $api->get('products/123456789');
echo sprintf('#%d. %s', $product['id'], $product['title']);
```

#### Update item
```php
$product = $api->put('products/123456789', [
    'title' => 'New title'
]);
echo sprintf('#%d. %s', $product['id'], $product['title']); // #1. New title
```

#### Create item
```php
$product = $api->post('products', [
    'title' => 'New product' 
]);
```

#### Delete item
```php
$api->delete('products/123456789');
if ($api->respCode == 200) {
    // Item has been successfully removed
}
```

Collection
-
You can use Collection object to get all the items from the specific endpoint. 
This library works fine with both page-based and cursor-based pagination and switches between them based on API version.
```php
use VladimirCatrici\Shopify\ClientManager;
$api = ClientManager::get('default');
$products = new Collection($api, 'products');
foreach ($products as $product) {
    printf('#%d. %s [$%f], 
        $product['id'], $product['title'], $product['price']
    );
}
```

Webhooks
-
You can use this library to listen for shop's webhooks.
```php
use VladimirCatrici\Shopify;
if (Shopify\Webhook::validate('your-webhook-token')) {
    printf('`%s` webhook triggered on your store (%s). Data received: %s', 
        Shopify\Webhook::getTopic(), 
        Shopify\Webhook::getShopDomain(),
        Shopify\Webhook::getData()
    );
} else {
    // Invalid request | Unauthorized webhook | Data corrupted
}

// You can also get webhook data as array right away
$data = Shopify\Webhook::getDataAsArray();
printf('Product ID#%d, product title: %s', 
    $data['id'], $data['title']
);
```

Troubleshooting
-
```php
use VladimirCatrici\Shopify\Exception\RequestException;
try {
    $products = $api->get('products');
} catch (RequestException $e) {
    $request = $e->getRequest(); // PSR-7/Request object

    $response = $e->getResponse(); // PSR-7/Response object
    $code = $response->getStatusCode(); // int
    $headers = $response->getHeaders(); // array
    $bodyStream = $response->getBody(); // Stream (PSR-7/StreamInterface)
    $bodyContent = (string) $response->getBody(); // string (or $body->__toString())
    
    // Details of the errors including exception message, request and response details
    echo $e->getDetailsJson();
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.2.5 - 2020-01-13
### Fixed
-   Use $_SERVER instead of filter_input(INPUT_SERVER, ...) as not working in PHP versions with enabled FastCGI

## 0.2.4 - 2020-01-13
### Added
-   Webhook::getHmacSha256() method visibility changed from private to public

## 0.2.3 - 2019-12-04
### Added
-   PHPDoc for ClientInterface, it was unclear before that client method may throw exceptions

## 0.2.2 - 2019-12-03
### Added
-   Allow `.myshopify.com` to be included in shop's handle passed to ClientConfig

### Fixed
-   Throw Guzzle request exception right away if it doesn't have response object to handle

### Removed
-   `vlucas/phpdotenv` dependency, moved to `require-dev`

## 0.2.1 - 2019-11-24
### Fixed
-   Call to a member function on null fix in RequestException
-   Autoload file with functions

## 0.2.0 - 2019-11-22
### Added
-   Client class added as a replacement of the API class 
-   ClientConfig added to configure new Shopify API client
-   Option to set up response data formatter. ResponseDefaultFormatter added and used by default
-   Declaring strict types

### Deprecated
-   API class deprecated, use Client instead

## 0.1.4 - 2019-11-24
### Fixed
-   Call to a member function on null fix in RequestException

## 0.1.3 - 2019-11-19
### Added
-   RequestException->getRequest() method added

## 0.1.2 - 2019-11-18
### Fixed
-   RequestException->getDetailsJson() method now returns correct request details

## 0.1.1 - 2019-11-14
### Added
-   CHANGELOG.md

## 0.1.0 - 2019-11-04
### Added
-   Initial release


src/API.php
<?php

namespace VladimirCatrici\Shopify;

use Exception;
use VladimirCatrici\Shopify\Exception\RequestException;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Response;
use InvalidArgumentException;
use Psr\Http\Message\StreamInterface;
use VladimirCatrici\Shopify\Response\ResponseDefaultFormatter;
use VladimirCatrici\Shopify\Response\ResponseDataFormatterInterface;

/**
 * Class API
 * @package VladimirCatrici\Shopify
 * @deprecated This class is deprecated since v0.2.0 and will be removed in v0.3.0. Use Client class instead
 */
class API implements ClientInterface
{
    /**
     * @var string Shopify store handle. If you shop permanent domain is test.myshopify.com - the "test" is the handle
     * in this case
     */
    private $handle;
    /**
     * @var string Full URL to the API including protocol and api API version e.g.
     * https://test.myshopify.com/admin/api/2019-10/
     */
    private $baseUrl;
    /**
     * @var string
     */
    private $accessToken;
    /**
     * @var integer Last request response code
     */
    public $respCode;
    /**
     * @var array Last request response headers
     */
    public $respHeaders;
    /**
     * @var Client
     */
    private $client;
    /**
     * @var ResponseDataFormatterInterface
     */
    private $responseFormatter;
    private $options = [
        /**
         * @var int Number of attempts trying to execute the request.
         * It's useful because sometimes Shopify may respond with 500 error.
         * The recommended value is 2 or 3
         */
        'max_attempts_on_server_errors' => 1,

        /**
         * @var int Number of attempts trying to execute the request on getting 429 Too Many Connections error.
         * This might be useful if the same API key is used by other apps which may lead to exceeding the rate limit.
         * The recommended value would be somewhere between would be up to 10
         */
        'max_attempts_on_rate_limit_errors' => 1,

        /**
         * @var float Number between 0 and 1 describing the maximum limit rate the client should reach before going
         * sleep. See `max_limit_rate_sleep_sec` option
         */
        'max_limit_rate' => 0.5,

        /**
         * @var int Number of seconds to sleep when API reaches the maximum API limit rate specified in `max_limit_rate`
         * option
         */
        'max_limit_rate_sleep_sec' => 1,

        /**
         * @var string API version, expecting format is YYYY-MM. By default uses the oldest supported stable version.
         */
        'api_version' => null,

        /**
         * @var string Response data formatter class name
         */
        'response_data_formatter' => null
    ];

    /**
     * API constructor.
     * @param string $domain Shopify domain
     * @param string $accessToken
     * @param array $clientOptions GuzzleHttp client options
     */
    public function __construct($domain, $accessToken, $clientOptions = [])
    {
        $this->handle = trim(preg_replace('/\.myshopify\.com$/', '', $domain));
        $this->accessToken = $accessToken;

        $internalOptions = array_keys($this->options);
        foreach ($internalOptions as $optionName) {
            if (array_key_exists($optionName, $clientOptions)) {
                $this->setOption($optionName, $clientOptions[$optionName]);
                unset($clientOptions[$optionName]);
            }
        }

        $this->baseUrl = 'https://' . $this->handle . '.myshopify.com/admin/' . (
            !empty($this->options['api_version']) ? 'api/' . $this->options['api_version'] . '/' : '');
        $client = new Client([
                'base_uri' => $this->baseUrl,
                'headers' => [
                    'Content-Type' => 'application/json',
                    'X-Shopify-Access-Token' => $this->accessToken
                ]
            ] + $clientOptions);
        $this->setClient($client);

        // Initialize default response data formatter (if necessary)
        if (empty($this->responseFormatter)) {
            $this->responseFormatter = new ResponseDefaultFormatter();
        }
    }

    /**
     * @param Client $client
     */
    private function setClient($client)
    {
        $this->client = $client;
    }

    /**
     * @param string $endpoint
     * @param array $query
     * @return mixed|StreamInterface
     * @throws RequestException
     */
    public function get($endpoint, $query = [])
    {
        return $this->request('get', $endpoint, $query);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed|StreamInterface
     * @throws RequestException
     */
    public function post($endpoint, $data = [])
    {
        return $this->request('post', $endpoint, null, $data);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed|StreamInterface
     * @throws RequestException
     */
    public function put($endpoint, $data = [])
    {
        return $this->request('put', $endpoint, null, $data);
    }

    /**
     * @param string $endpoint
     * @return mixed|StreamInterface
     * @throws RequestException
     */
    public function delete($endpoint)
    {
        return $this->request('delete', $endpoint);
    }

    /**
     * @noinspection PhpDocMissingThrowsInspection
     * @param $method
     * @param $endpoint
     * @param array $query
     * @param array $data
     * @return mixed|StreamInterface
     * @throws RequestException
     */
    private function request($method, $endpoint, $query = [], $data = [])
    {
        $fullApiRequestURL = $this->generateFullApiRequestURL($endpoint, $query);

        $maxAttemptsOnServerErrors = $this->getOption('max_attempts_on_server_errors');
        $maxAttemptsOnRateLimitErrors = $this->getOption('max_attempts_on_rate_limit_errors');

        $serverErrors = 0;
        $rateLimitErrors = 0;
        $lastException = null;
        $options = [];
        if (in_array($method, ['put', 'post']) && !empty($data)) {
            $options = [
                'json' => $data
            ];
        }
        while ($serverErrors < $maxAttemptsOnServerErrors && $rateLimitErrors < $maxAttemptsOnRateLimitErrors) {
            try {
                /** @noinspection PhpUnhandledExceptionInspection */
                $response = $this->client->request($method, $fullApiRequestURL, $options);
                break;
            } catch (\GuzzleHttp\Exception\RequestException $e) {
                $lastException = new RequestException($this->client, $e);
                $handlerResult = $this->handleRequestException($e);
                if ($handlerResult['break']) {
                    break;
                }
                $serverErrors += $handlerResult['server_error'];
                $rateLimitErrors += $handlerResult['rate_limit_error'];
            }
        }

        if (empty($response)) {
            throw $lastException;
        }

        $this->respCode = $response->getStatusCode();
        $this->respHeaders = $response->getHeaders();
        $this->rateLimitSleepIfNeeded($response);

        // Response Body
        $body = $response->getBody()->getContents();
        return $this->responseFormatter->output($body);
    }

    /**
     * @param $guzzleRequestException \GuzzleHttp\Exception\RequestException
     * @return array
     */
    private function handleRequestException($guzzleRequestException)
    {
        $output = [
            'server_error' => 0,
            'rate_limit_error' => 0,
            'break' => false
        ];
        switch ($guzzleRequestException->getResponse()->getStatusCode()) {
            case 500:
            case 503:
            case 504:
                $output['server_error'] = 1;
                break;
            case 429:
                $output['rate_limit_error'] = 1;
                if ($guzzleRequestException->getResponse()->hasHeader('Retry-After')) {
                    sleep($guzzleRequestException->getResponse()->getHeaderLine('Retry-After'));
                    break;
                }
                sleep(1);
                break;
            default:
                $output['break'] = true;
        }
        return $output;
    }

    private function generateFullApiRequestURL($endpoint, $queryParams = [])
    {
        if (!preg_match('/\.json$/', $endpoint)) {
            $endpoint .= '.json';
        }
        if (!empty($queryParams)) {
            $queryString = http_build_query($queryParams);
            return $endpoint . '?' . $queryString;
        }
        return $endpoint;
    }

    private function rateLimitSleepIfNeeded(Response $response)
    {
        if ($response->hasHeader('X-Shopify-Shop-Api-Call-Limit')) {
            $limit = explode('/', $response->getHeaderLine('X-Shopify-Shop-Api-Call-Limit'));
            $rate = $limit[0] / $limit[1];
            if ($rate > $this->getOption('max_limit_rate')) {
                sleep($this->getOption('max_limit_rate_sleep_sec'));
            }
        }
    }

    public function getOption($key)
    {
        if (!array_key_exists($key, $this->options)) {
            throw new InvalidArgumentException(sprintf('Trying to get an invalid option `%s`', $key));
        }
        return $this->options[$key];
    }

    public function setOption($key, $value)
    {
        if (!array_key_exists($key, $this->options)) {
            throw new InvalidArgumentException(sprintf('Trying to set an invalid option `%s`', $key));
        }

        if ($key == 'api_version') {
            $value = trim($value);
            $this->validateApiVersion($value);

            // Update client config
            if (!empty($this->client)) {
                $currentClientConfig = $this->client->getConfig();
                $this->baseUrl = preg_replace(
                    '/admin\/(api\/\d{4}-\d{2}\/)?$/',
                    'admin/api/' . $value . '/',
                    $this->baseUrl
                );
                $currentClientConfig['base_uri'] = $this->baseUrl;
                $this->setClient(new Client($currentClientConfig));
            }
        }

        if ($key == 'response_data_formatter') {
            $this->responseFormatter = new $value();
        }

        $this->options[$key] = $value;
    }

    private function validateApiVersion($value)
    {
        $readMoreAboutApiVersioning = 'Read more about versioning here: https://help.shopify.com/en/api/versioning';
        if (!preg_match('/^(\d{4})-(\d{2})$/', $value, $matches)) {
            throw new InvalidArgumentException(
                sprintf(
                    'Invalid API version format: "%s". The "YYYY-MM" format expected. ' . $readMoreAboutApiVersioning,
                    $value
                )
            );
        }
        if ($matches[1] < 2019) {
            throw new InvalidArgumentException(
                sprintf(
                    'Invalid API version year: "%s". The API versioning has been released in 2019. ' . $readMoreAboutApiVersioning,
                    $value
                )
            );
        }
        if (!preg_match('/01|04|07|10/', $matches[2])) {
            throw new InvalidArgumentException(
                sprintf(
                    'Invalid API version month: "%s". 
                    The API versioning has been released on April 2019 and new releases scheduled every 3 months, 
                    so only "01", "04", "07" and "10" expected as a month. Otherwise, 
                    "404 Not Found" will be returned by Shopify. ' . $readMoreAboutApiVersioning,
                    $matches[0]
                )
            );
        }
    }

    public function setVersion($version)
    {
        $this->setOption('api_version', $version);
    }

    /**
     * @return string
     * @throws Exception
     */
    public function getVersion()
    {
        $version = $this->getOption('api_version');
        if (empty($version)) {
            $headerKey = 'X-Shopify-API-Version';
            if (is_array($this->respHeaders) && array_key_exists($headerKey, $this->respHeaders)) {
                $this->options['api_version'] = $this->respHeaders[$headerKey][0];
            } else {
                $this->options['api_version'] = static::getOldestSupportedVersion();
            }
        }
        return $this->getOption('api_version');
    }

    /**
     * @param mixed $date DateTime string (YYYY-MM[...]) or DateTime object
     * @return string
     * @throws Exception
     */
    public static function getOldestSupportedVersion($date = '')
    {
        return getOldestSupportedVersion($date);
    }
}
